### ISSUE TITLE
Issue with weighting algo in a stack.

### ISSUE THREAD

**Issue by nikolas-joyce** [pmorissette/bt#150]
I am trying to combine a set of back test outcomes into a new portfolio and i am having difficulty using anything but the default setting in a few of the algos.

here a code snippet..

`<
data = bt.merge(res['corn_u_mo'].prices,
                res['tlt_u_mo'].prices,
                res['uso_u_mo'].prices,
                res['ung_u_mo'].prices,
                res['cane_u_mo'].prices,
                res['jo_u_mo'].prices,
                res['slv_u_mo'].prices,
                res['gld_u_mo'].prices,
                res['soyb_u_mo'].prices,
                res['uco_u_mo'].prices,
                res['pplt_u_mo'].prices,
                res['dbb_u_mo'].prices,
               )

res = bt.run(t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, t11, t12)

 long_sma  = data.rolling(window=5,center=False).median()

tw_P = long_sma.copy()
tw_P[data > long_sma] = 1.0  
tw_P[data <= long_sma] = 0.0
tw_P[long_sma.isnull()] = 0.0

universal_momentum_index_strategy_InvVol = bt.Strategy('universal_momentum_index_InvVol', 
                                                                   [bt.algos.SelectAll(),
                                                                     bt.algos.WeighTarget(tw_P),
                                                                    bt.algos.WeighEqually(), 
                                                                    bt.algos.SelectMomentum(n = 2, lookback=pd.DateOffset(months=2)),
                                                                    bt.algos.WeighInvVol(),
                                                                    bt.algos.LimitWeights(limit=0.4), 
                                                                    bt.algos.Rebalance()])
>`

######

Basically i would like to see some none default or blank values in the stack of algos but i cannot get the formatting correct.
bt.algos.SelectMomentum(n = 2, lookback=pd.DateOffset(months=2)), should this work ?

what can i but between the () in this invol algo?
bt.algos.WeighInvVol(),

![image](https://user-images.githubusercontent.com/18488108/40462674-ce5912f8-5ec5-11e8-8be6-a6b03c40d4aa.png)

this works my i am unsure if the order of the algos needs to be in a certain order. 
any advice.
thx
ps you can get the notebook here.

https://github.com/nikolas-joyce/Universal_Momentum
thx

**Comment by JordanPlatts** [comment]
t0 = target.now - self.lag
    prc = target.universe[selected].loc[t0 - self.lookback:t0]

Lag `k` sets the time back from `t` to `t-k`. This would be helpful if you were looking at data that gets revised and you want to wait till the revised number comes out so you can act on it. For example, today the Initial Jobless Claims from last week was revised to 223k. So to incorporate this data you could lag the Initial Jobless Claims by 1 week.

To set the lookback to 6 months and to include today's prices use

    bt.algos.WeighInvVol(lookback=pd.DateOffset(months=6),lag=pd.DateOffset(days=0))

**Comment by nikolas-joyce** [comment]
thx that looks good.

One question regarding bt.algos.WeighTarget(tw_P). Is it clear why the following code would fail to have target weights influence the outcome?

`data = bt.merge(res['corn_u_mo'].prices,
                res['tlt_u_mo'].prices,
                res['uso_u_mo'].prices,
                res['ung_u_mo'].prices,
                res['cane_u_mo'].prices,
                res['jo_u_mo'].prices,
                res['slv_u_mo'].prices,
                res['gld_u_mo'].prices,
                res['soyb_u_mo'].prices,
                res['uco_u_mo'].prices,
                res['pplt_u_mo'].prices,
                res['dbb_u_mo'].prices,
               )


res = bt.run(t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, t11, t12)

long_sma  = data.rolling(window=5,center=False).median()

tw_P = long_sma.copy()
tw_P[data > long_sma] = 1.5  #this weighting does not seem to be working any advice?
tw_P[data <= long_sma] = 0.5
tw_P[long_sma.isnull()] = 0.0


universal_momentum_index_strategy_InvVol = bt.Strategy('universal_momentum_index_InvVol', [bt.algos.SelectAll(),
                                                                    bt.algos.WeighTarget(tw_P),
                                                                    bt.algos.WeighInvVol(lookback=pd.DateOffset(months=3),lag=pd.DateOffset(days=0)),
                                                                    bt.algos.LimitWeights(limit=0.7), 
                                                                    bt.algos.Rebalance()])`

for some reason the target weights do not seem to be working on the Strategy.
No matter what values i choose i am getting the same result. tw_P[data > long_sma] = 1.5  or tw_P[data > long_sma] = 0 give the same outcome? 
Should i be calling bt.algos.WeighTarget(tw_P), at a different place in the stack?
thx

**Comment by JordanPlatts** [comment]
Your strategy is doing the following:

1. You don't have a Run*(RunDaily,RunMonthly, etc.) algo to start with so it runs on every datetime in the dataframe.
2. bt.algos.SelectAll() then saves all of the securities with prices above 0 to target.temp['selected']
3. bt.algos.WeighTarget(tw_P) then saves the weights not with NA values in target.temp['weights'] = w.dropna(). Note this doesn't use anything from SelectAll().
4. bt.algos.WeighInvVol(lookback=pd.DateOffset(months=3),lag=pd.DateOffset(days=0)) then takes the securities from target.temp['selected'](set by SelectAll()) and overwrites target.temp['weights'] to have their weights have equal volatility.
5. bt.algos.LimitWeights(limit=0.7) then goes through target.temp['weights'] and takes any weight > 70% and redistributes it proportionally to the other securities in target.temp['weights']
6. bt.algos.Rebalance() then takes the weights in target.temp['weights'] and transacts to change the portfolio to match the desired weights.

In summary, this error is cause because you have both bt.algos.WeighTarget(tw_P) and bt.algos.WeighInvVol(lookback=pd.DateOffset(months=3),lag=pd.DateOffset(days=0)) running.

**Comment by JordanPlatts** [comment]
To be technical, WeighInvVol is just overwriting WeighTarget because it comes after.

If you are going to use this library a lot, I would recommend using pycharm because it has a debugger which helps a lot to see what the code is doing.

**Comment by nikolas-joyce** [comment]
Great thx for the suggestion.

Sent from my iPhone

On May 24, 2018, at 2:28 PM, Jordan Platts <notifications@github.com<mailto:notifications@github.com>> wrote:


To be technical, WeighInvVol is just overwriting WeighTarget because it comes after.

If you are going to use this library a lot, I would recommend using pycharm because it has a debugger which helps a lot to see what the code is doing.
